modules/contact/src/userobjects/PenaltyFrictionUserObject.C

```diff
@@ -0,0 +1,351 @@
+  return params;
+}
+
+PenaltyFrictionUserObject::PenaltyFrictionUserObject(const InputParameters & parameters)
+  : WeightedVelocitiesUserObject(parameters),
+    _penalty(getParam<Real>("penalty")),
+    _penalty_friction(isParamValid("penalty_friction") ? getParam<Real>("penalty_friction")
+                                                       : getParam<Real>("penalty")),
+    _friction_coefficient(getParam<Real>("friction_coefficient"))
+{
+  auto check_type = [this](const auto & var, const auto & var_name)
+  {
+    if (!var.isNodal())
+      paramError(var_name,
+                 "The displacement variables must have degrees of freedom exclusively on "
+                 "nodes, e.g. they should probably be of finite element type 'Lagrange'.");
+  };
+  check_type(*_disp_x_var, "disp_x");
+  check_type(*_disp_y_var, "disp_y");
+  if (_has_disp_z)
+    check_type(*_disp_z_var, "disp_z");
+}
+
+const VariableTestValue &
+PenaltyFrictionUserObject::test() const
+{
+  return _disp_x_var->phiLower();
+}
+
+const ADVariableValue &
+PenaltyFrictionUserObject::contactPressure() const
+{
+  return _contact_force;
+}
+
+const ADVariableValue &
+PenaltyFrictionUserObject::contactTangentialPressureDirOne() const
+{
+  return _frictional_contact_force_one;
+}
+
+const ADVariableValue &
+PenaltyFrictionUserObject::contactTangentialPressureDirTwo() const
+{
+  return _frictional_contact_force_two;
+}
+
+void
+PenaltyFrictionUserObject::timestepSetup()
+{
+  WeightedVelocitiesUserObject::timestepSetup();
+
+  _dof_to_old_weighted_gap.clear();
+  _dof_to_old_real_tangential_velocity.clear();
+  _dof_to_old_accumulated_slip.clear();
+  _dof_to_old_normal_pressure.clear();
+
+  for (auto & map_pr : _dof_to_weighted_gap)
+    _dof_to_old_weighted_gap.emplace(map_pr.first, std::move(map_pr.second.first));
+
+  for (auto & map_pr : _dof_to_real_tangential_velocity)
+    _dof_to_old_real_tangential_velocity.emplace(map_pr);
+
+  for (auto & map_pr : _dof_to_normal_pressure)
+    _dof_to_old_normal_pressure.emplace(map_pr.first, MetaPhysicL::raw_value(map_pr.second));
+
+  for (auto & map_pr : _dof_to_accumulated_slip)
+  {
+    if (_dof_to_old_normal_pressure[map_pr.first] >
+        TOLERANCE * TOLERANCE) // Only if it had normal pressure. Otherwise, restart count
+      _dof_to_old_accumulated_slip.emplace(map_pr);
+    else
+      _dof_to_old_accumulated_slip[map_pr.first] = {0.0, 0.0};
+  }
+}
+
+void
+PenaltyFrictionUserObject::initialize()
+{
+  WeightedVelocitiesUserObject::initialize();
+
+  _dof_to_normal_pressure.clear();
+  _dof_to_accumulated_slip.clear();
+  _dof_to_frictional_pressure.clear();
+}
+
+Real
+PenaltyFrictionUserObject::getNormalContactPressure(const Node * const node) const
+{
+  for (auto & map_pr : _dof_to_normal_pressure)
+    if (map_pr.first->id() == node->id())
+      return MetaPhysicL::raw_value(
+          libmesh_map_find(_dof_to_normal_pressure, static_cast<const DofObject *>(map_pr.first)));
```

Comment:
```suggestion
  if (auto it = _dof_to_normal_pressure.find(_mesh.nodePtr(node->id()));
      it != _dof_to_normal_pressure.end())
    return MetaPhysicL::raw_value(it->second);
```
Let's not do a linear search

Reply:
All these comments on not doing linear search should be addressed in the latests commit.